src/lib/y2storage/autoinst_proposal.rb

```diff
@@ -0,0 +1,184 @@
+    attr_reader :initial_devicegraph
+
+    # Proposed layout of devices, nil if the proposal has not been
+    # calculated yet
+    # @return [Devicegraph]
+    attr_reader :proposed_devicegraph
+
+    alias_method :devices, :proposed_devicegraph
+
+    # Constructor
+    #
+    # @param partitioning [Array<Hash>] Partitioning schema from an AutoYaST profile
+    # @param devicegraph  [Devicegraph] starting point. If nil, then probed devicegraph
+    #   will be used
+    # @param disk_analyzer [DiskAnalyzer] if nil, a new one will be created
+    #   based in the initial devicegraph
+    def initialize(partitioning: [], devicegraph: nil, disk_analyzer: nil)
+      @partitioning = partitioning
+      @initial_devicegraph = devicegraph
+      @disk_analyzer = disk_analyzer
+      @proposed = false
+    end
+
+    # Checks whether the proposal has already been calculated
+    #
+    # @return [Boolean]
+    def proposed?
+      @proposed
+    end
+
+    # Calculates the proposal
+    #
+    # @raise [UnexpectedCallError] if called more than once
+    # @raise [NoDiskSpaceError] if there is no enough space to perform the installation
+    def propose
+      raise UnexpectedCallError if proposed?
+
+      drives = Proposal::AutoinstDrivesMap.new(initial_devicegraph, partitioning)
+
+      space_maker = Proposal::AutoinstSpaceMaker.new(disk_analyzer)
+      devicegraph = space_maker.provide_space(initial_devicegraph, drives)
+
+      @proposed_devicegraph = propose_devicegraph(devicegraph, drives)
+      @proposed = true
+
+      nil
+    end
+
+  protected
+
+    # Proposes a devicegraph based on given drives map
+    #
+    # This method falls back to #proposed_guided_devicegraph when the device map
+    # does not contain any partition.
+    #
+    # @param devicegraph [Devicegraph]                 Starting point
+    # @param drives      [Proposal::AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Devicegraph containing the planned devices
+    def propose_devicegraph(devicegraph, drives)
+      planned_devices = plan_devices(devicegraph, drives)
+      create_devices(devicegraph, planned_devices, drives.disk_names)
+    rescue Error
+      log.info "No partitions were specified. Falling back to guided setup planning."
+      propose_guided_devicegraph(devicegraph, drives)
+    end
+
+    # Finds a suitable devicegraph using the guided proposal approach
+    #
+    # If the :desired target fails, it will retry with the :min target.
+    #
+    # @param devicegraph [Devicegraph]       Starting point
+    # @param drives      [AutoinstDrivesMap] Devices map from an AutoYaST profile
+    # @return [Devicegraph] Proposed devicegraph using the guidede proposal approach
+    #
+    # @raise [Error] No suitable devicegraph was found
+    # @see proposed_guided_devicegraph
+    def propose_guided_devicegraph(devicegraph, drives)
+      guided_devicegraph_for_target(devicegraph, drives, :desired)
+    rescue Error
```

Comment:
It's about retrying, so we will retry with a different target on any Storage error.